src/main/java/redis/clients/jedis/JedisFactory.java

```diff
@@ -81,21 +73,20 @@
         "Cannot open Redis connection due invalid URI. %s", uri.toString()));
     }
 
-    this.hostAndPort.set(new HostAndPort(uri.getHost(), uri.getPort()));
-    this.connectionTimeout = connectionTimeout;
-    this.soTimeout = soTimeout;
+    this.jedisSocketFactory = new DefaultJedisSocketFactory(
+        uri.getHost(), uri.getPort(), connectionTimeout, soTimeout,
+        JedisURIHelper.isRedisSSLScheme(uri), sslSocketFactory,
+        sslParameters, hostnameVerifier
+    );
     this.user = JedisURIHelper.getUser(uri);
     this.password = JedisURIHelper.getPassword(uri);
     this.database = JedisURIHelper.getDBIndex(uri);
     this.clientName = clientName;
-    this.ssl = JedisURIHelper.isRedisSSLScheme(uri);
-    this.sslSocketFactory = sslSocketFactory;
-    this.sslParameters = sslParameters;
-    this.hostnameVerifier = hostnameVerifier;
   }
 
   public void setHostAndPort(final HostAndPort hostAndPort) {
```

Comment:
This should be `synchronized` for reliablity as there is no AtomicReference.

Reply:
That's a good point, but even if we add `synchronized` it would still be prone to some corner cases in the other methods since make/validate can still happen mid-updating the host/port.

I can think of a few solutions here (no particular ordering beside preference towards first one):
* We put the jedis socket factory in an atomic reference, we make the setHostAndPort copy/update/reset the jedis socket factory
* We put `synchronized` on all methods (overkill and I really think that's a bad option, but just listing)
* We keep an instance of HostAndPort in an atomic reference and cross check/validate with that (but then we duplicate the state and make it a bit too complex)
* The `setHostAndPort` method is only used in [sentinel pool](https://github.com/redis/jedis/blob/f24a817987fe1990a9ff65b31dba7fa4b44969e6/src/main/java/redis/clients/jedis/JedisSentinelPool.java#L180), we can mark the method as deprecated with a comment for removal in a future release, and change the sentinel bit to re-create the whole pool (although I don't know the sentinel code that well so maybe it's a bad idea)

What do you think?

Reply:
Hmm, first one does seem like a prominent one if you have actually meant copy+update+reset.

Reply:
Yup, I meant all not either (shouldn't have used `/` here :D), so something like:

```java
JedisSocketFactory newFactory = jedisSocketFactory.get().clone();
newFactory.setHost(...);
newFactory.setPort(...);
jedisSocketFactory.set(newFactory);
```